x-pack/plugin/autoscaling/src/main/java/org/elasticsearch/xpack/autoscaling/master/DedicatedMasterNodesDeciderService.java

```diff
@@ -0,0 +1,172 @@
+import org.elasticsearch.xpack.autoscaling.capacity.AutoscalingCapacity;
+import org.elasticsearch.xpack.autoscaling.capacity.AutoscalingDeciderContext;
+import org.elasticsearch.xpack.autoscaling.capacity.AutoscalingDeciderResult;
+import org.elasticsearch.xpack.autoscaling.capacity.AutoscalingDeciderService;
+
+import java.io.IOException;
+import java.util.List;
+import java.util.Objects;
+import java.util.Set;
+import java.util.stream.StreamSupport;
+
+public class DedicatedMasterNodesDeciderService implements AutoscalingDeciderService {
+    public static final String NAME = "dedicated-masters-decider";
+
+    public static final int DEFAULT_NUMBER_OF_MASTER_NODES = 3;
+    public static final ByteSizeValue DEFAULT_MASTER_MEMORY = ByteSizeValue.ofGb(1);
+    public static final int DEFAULT_MIN_HOT_NODES = 6;
+    public static final ByteSizeValue DEFAULT_HOT_NODE_MIN_MEMORY = ByteSizeValue.ofGb(64);
+
+    public static final Setting<ByteSizeValue> MASTER_NODE_MEMORY = Setting.byteSizeSetting(
+        "master_node_memory",
+        (ignored) -> DEFAULT_MASTER_MEMORY.getStringRep(),
+        DEFAULT_MASTER_MEMORY,
+        ByteSizeValue.ofBytes(Long.MAX_VALUE)
+    );
+    public static final Setting<Integer> MIN_HOT_NODES = Setting.intSetting("min_hot_nodes", DEFAULT_MIN_HOT_NODES, 1);
+    public static final Setting<ByteSizeValue> MIN_HOT_NODE_MEMORY = Setting.byteSizeSetting(
+        "min_hot_node_memory",
+        (ignored) -> DEFAULT_HOT_NODE_MIN_MEMORY.getStringRep(),
+        ByteSizeValue.ofGb(1),
+        ByteSizeValue.ofBytes(Long.MAX_VALUE)
+    );
+
+    @Override
+    public AutoscalingDeciderResult scale(Settings configuration, AutoscalingDeciderContext context) {
+        int hotAndContentNodes = (int) StreamSupport.stream(context.state().nodes().spliterator(), false)
+            .filter(this::hasHotAndContentRole)
+            .count();
+        long totalHotAndContentNodesMemory = StreamSupport.stream(context.state().nodes().spliterator(), false)
+            .filter(this::hasHotAndContentRole)
+            .mapToLong(node -> getNodeMemory(node, context))
+            .sum();
+
+        final int minHotNodes = MIN_HOT_NODES.get(configuration);
+        final ByteSizeValue minHotNodeMemory = MIN_HOT_NODE_MEMORY.get(configuration);
+        final ByteSizeValue masterNodeMemory = MASTER_NODE_MEMORY.get(configuration);
+        final AutoscalingCapacity autoscalingCapacity;
+        if (hotAndContentNodes >= minHotNodes && totalHotAndContentNodesMemory >= minHotNodes * minHotNodeMemory.getBytes()) {
+            autoscalingCapacity = AutoscalingCapacity.builder()
+                .node(null, masterNodeMemory)
+                .total(null, ByteSizeValue.ofBytes(DEFAULT_NUMBER_OF_MASTER_NODES * masterNodeMemory.getBytes()))
+                .build();
+        } else {
+            autoscalingCapacity = AutoscalingCapacity.ZERO;
+        }
+
+        return new AutoscalingDeciderResult(
```

Comment:
Repeating our conclusion from another channel. For now, I think we should only respond with `current_capacity` or higher to ensure we do not signal a scale down. This is needed to avoid that the loss of a single node causes a reduction in master size too (or signaling to switch back to not having dedicated masters).

Once desired nodes is in place, we can remove this again.